### DISCUSSION TITLE
Question about multiply and divide

### DISCUSSION THREAD

**Discussion by rebornfz** [libvips/libvips#2229] [Q&A] [answered]
For example,
vips::VImage  image = image.multiply(image_input.new_from_image({1})) ;

vips::VImage  image = image.divide(image_input.new_from_image({1})) ;

The above get different results

Why？

The image_input, an one band image
![1](https://user-images.githubusercontent.com/27042311/116395974-5d8bdb80-a857-11eb-8ee6-122cef249e97.png)

**Comment by rebornfz** [comment]
I want to change the transparency of the picture.
I separate the alpha band, multiply it by the opacity(0-1), and then merge the bands, like this
`image_op_backgroung = image_input.extract_band(3);`
`image_input = image_input.extract_band(0,vips::VImage::option()->set("n", 3));`
`image_op_backgroung = image_op_backgroung.multiply(image_input.new_from_image({opacity}));`
`image_input = image_input.bandjoin(image_op_backgroung);`
But it went wrong

If I do like this
`image_op_backgroung = image_input.extract_band(3);`
`image_input = image_input.extract_band(0,vips::VImage::option()->set("n", 3));`
`image_op_backgroung = image_op_backgroung.divide(image_input.new_from_image({1 / opacity}));`
`image_input = image_input.bandjoin(image_op_backgroung);`
It worked well

Why?

**Reply by jcupitt** [reply]
You can use a vector for the constant, eg.:

```
a *= [1, 1, 1, 0.5]
```

No need to split and recombine.

**Reply by jcupitt** [reply]
Or in C++:

```C++
image_input *= std::vector<double>{1, 1, 1, opacity};
```

**Comment by jcupitt** [comment]
It should be the same. With your test image I see:

```
john@banana ~/pics $ python3
Python 3.8.6 (default, Jan 27 2021, 15:42:20) 
[GCC 10.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyvips
>>> a = pyvips.Image.new_from_file("test.png")
>>> b = a.new_from_image(1)
>>> (a * b).avg()
17.674368489583333
>>> (a / b).avg()
17.674368489583333
>>> ((a * b) == (a / b)).min()
255.0
>>> ((a * b) == a).min()
255.0
>>> ((a / b) == a).min()
255.0
>>> 
```

**Comment by rebornfz** [comment]
It my result
mutiply:
![m](https://user-images.githubusercontent.com/27042311/116398699-ad1fd680-a85a-11eb-93e2-74b664e40e06.png)
divide:
![d](https://user-images.githubusercontent.com/27042311/116398733-b446e480-a85a-11eb-8a4a-13ef749165da.png)
The avg is same, but write to file is different.
Do I need to do something before writing to file?
@jcupitt

**Reply by jcupitt** [reply]
Perhaps you are saving them to a different image format? Perhaps you are using a very old libvips with a strange bug? Perhaps you have an error somewhere elsewhere in your code? I don't know.

You need to make a complete program that I can run that demonstrates the problem.

**Comment by rebornfz** [comment]
My code:
```
#include <iostream>
#include <vips/vips8>

using namespace vips;
using namespace std;

int
main (int argc, char **argv)
{
        if (VIPS_INIT (argv[0]))
                vips_error_exit (NULL);
        vips::VImage image_op_backgroung;
        vips::VImage image_op;
        vips::VImage image = vips::VImage::new_from_file("test.png");
        if(image.has_alpha())
        {
            image_op_backgroung = image.extract_band(3);
            image = image.extract_band(0,
                vips::VImage::option()->set("n", 3));
        }
        image_op = image_op_backgroung.multiply(image.new_from_image({1}));
        std::cout << "avg " << image_op_backgroung.avg() << std::endl;
        image_op.write_to_file("m.png");
        image_op = image_op_backgroung.divide(image.new_from_image({1}));
        std::cout << "avg " << image_op_backgroung.avg() << std::endl;
        image_op.write_to_file("d.png");

        return 0;
}
```

The test.png
![fe18323e3c70030d55eeed4cbfe34172](https://user-images.githubusercontent.com/27042311/116401395-c9714280-a85d-11eb-9a4e-3c8c4d6acdd3.png)

My result
multiply
![m](https://user-images.githubusercontent.com/27042311/116401466-dbeb7c00-a85d-11eb-9ca8-d1d37d8f1fb6.png)


divide:
![d](https://user-images.githubusercontent.com/27042311/116401535-eefe4c00-a85d-11eb-94ab-2c39f1a8782f.png)

avg 17.6744
avg 17.6744
@jcupitt

**Comment by jcupitt** [comment, marked as answer]
Ah, I see --- it's the numeric format.

Your image is 8-bit, so `.new_from_image({1})` will also make an 8-bit image.

When you multiply them, you'll get a 16-bit image, since it can represent two 8-bit images multiplied together perfectly. When you save as PNG, you'll get a 16-bit PNG, so the 0 - 255 in your image will look very dark. Multiply that image by 256 and you'll see your pixels are there.

When you divide two 8-bit images, you get a float image. When you save a float image as PNG, it's automatically converted to 8-bit for you, so you see the full range of brightnesses.

**Reply by rebornfz** [reply]
Thank you! Can I get a 8-bit PNG bymultiply with  two 8-bit images?

**Reply by jcupitt** [reply]
Sure, use `.cast()` to change the numeric type:

```C++
(a * b).cast(VIPS_FORMAT_UCHAR).write_to_file(argv[2]);
```

**Reply by rebornfz** [reply]
Or can I get a float image by multiply?

**Reply by rebornfz** [reply]
Thank you, I'll try it

**Reply by rebornfz** [reply]
It worked. I was confused for a long time, Thank you very much

**Comment by jcupitt** [comment]
You could write your program like this:

```C++
/* compile with
 *
 *      g++ -g -Wall try94.cpp `pkg-config vips-cpp --cflags --libs`
 *
 */ 

#include <vips/vips8>

using namespace vips;

int
main(int argc, char **argv)
{       
        if (VIPS_INIT(argv[0]))
                vips_error_exit(NULL);

        VImage a = VImage::new_from_file(argv[1]);
        VImage b = a.new_from_image({1});

        (a * b).write_to_file(argv[2]);
        (a / b).write_to_file(argv[3]);

        return 0;
}       
```

Or like this, of course:

```C++
#include <vips/vips8>

using namespace vips;

int
main(int argc, char **argv)
{       
        if (VIPS_INIT(argv[0]))
                vips_error_exit(NULL);

        VImage a = VImage::new_from_file(argv[1]);

        (a * 1.0).write_to_file(argv[2]);
        (a / 1.0).write_to_file(argv[3]);

        return 0;
}      
```

**Reply by rebornfz** [reply]
Thank you, It looks better

**Comment by jcupitt** [comment]
Don't forget, if you just want to change transparency, you can use a vector constant.

```C++
        VImage a = ...;

        (a * std::vector<double>{1, 1, 1, 0.5}).write_to_file(filename);
```
